Argo/Globals/StandardTypes.swift

```diff
@@ -2,53 +2,64 @@ import Foundation
 import Runes
 
 extension String: JSONDecodable {
-  public static func decode(j: JSON) -> String? {
+  public static func decode(j: JSON) -> Decoded<String> {
     switch j {
-    case let .String(s): return s
-    default: return .None
+    case let .String(s): return pure(s)
+    default: return typeMismatch("String", j)
     }
   }
 }
 
 extension Int: JSONDecodable {
-  public static func decode(j: JSON) -> Int? {
+  public static func decode(j: JSON) -> Decoded<Int> {
     switch j {
-    case let .Number(n): return n as Int
-    default: return .None
+    case let .Number(n): return pure(n as Int)
+    default: return typeMismatch("Int", j)
     }
   }
 }
 
 extension Double: JSONDecodable {
-  public static func decode(j: JSON) -> Double? {
+  public static func decode(j: JSON) -> Decoded<Double> {
     switch j {
-    case let .Number(n): return n as Double
-    default: return .None
+    case let .Number(n): return pure(n as Double)
+    default: return typeMismatch("Double", j)
     }
   }
 }
 
 extension Bool: JSONDecodable {
-  public static func decode(j: JSON) -> Bool? {
+  public static func decode(j: JSON) -> Decoded<Bool> {
     switch j {
-    case let .Number(n): return n as Bool
-    default: return .None
+    case let .Number(n): return pure(n as Bool)
+    default: return typeMismatch("Bool", j)
     }
   }
 }
 
 extension Float: JSONDecodable {
-  public static func decode(j: JSON) -> Float? {
+  public static func decode(j: JSON) -> Decoded<Float> {
     switch j {
-    case let .Number(n): return n as Float
-    default: return .None
+    case let .Number(n): return pure(n as Float)
+    default: return typeMismatch("Float", j)
     }
   }
 }
 
-public func decodeArray<A where A: JSONDecodable, A == A.DecodedType>(value: JSON) -> [A]? {
+public func decodeArray<A where A: JSONDecodable, A == A.DecodedType>(value: JSON) -> Decoded<[A]> {
   switch value {
-  case let .Array(a): return sequence(A.decode <^> a)
-  default: return .None
+  case let .Array(a): return sequence(a.map { A.decode($0) })
```

Comment:
Both of these can be `sequence(A.decode <^> a)` now. Might as well make that change while we're futzing around in here.